### PR TITLE
Increase effect of --dry-run flag (#7379)

### PR DESCRIPTION
**Pull request by aaronallen8455**
Changes the behavior of the following commands under the --dry-run flag
- `v2-configure` and `v2-freeze` no longer write files.
- `v2-exec` and `v2-run` print out the command that would've been
executed rather than doing so.

Adds package tests for the configure and freeze commands.

---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by aaronallen8455** [comment]
Tests fail because the output I added for freeze command is not agnostic to the version of ghc used i.e. the test runner produces base ==4.13 instead of base ==4.14.
Any thoughts on how to remedy this? Is this additional output even useful?

**Comment by fgaz** [comment]
> Any thoughts on how to remedy this?

Look at https://github.com/haskell/cabal/blob/master/cabal-testsuite/README.md, sections "How do I turn of accept tests? My test output wobbles to much" and "How can I manually test for a string in output?". But as you wrote

> Is this additional output even useful?

in this case it's probably better to just output "cabal.project.freeze would be written" or similar. The freeze file could be _very_ long after all, and its contents would just clutter the terminal

**Review by fgaz — APPROVED** [review]
Looks good!

I see that this only takes --dry-run into account, but that's fine, this is a net improvement and the --download-only check can be added later

**Comment by aaronallen8455** [comment]
@fgaz Thanks for reviewing. Should the `--only-download` flag have the same effect as `--dry-run` for all of these commands?

**Comment by fgaz** [comment]
Yes, since those are all steps that would be performed after the download. Do you plan to add it to this pr (just to know whether to wait for that before merging)?

**Comment by aaronallen8455** [comment]
I'll add it to this PR since it's a small change.

**Comment by fgaz** [comment]
Oh, I almost forgot, this needs a [changelog entry](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog) (looks good otherwise).

Finally, please rebase on current master. We're trying a github setting that requires prs to be updated on the base branch

**Review by fgaz — APPROVED** [review]
Thanks!

Pinging @Mikolaj @emilypi for the second review

**Review by emilypi — APPROVED** [review]

**Comment by fgaz** [comment]
@Mergifyio backport 3.6

**Comment by emilypi** [comment]
@aaronallen8455 this looks good, and I'd like to merge, but the recent v2-configure work looks like it caused conflicts. If you want, @ptkato is around to help sort those out.

**Comment by aaronallen8455** [comment]
I've resolved the conflict. Perhaps someone could take a look one more time.

**Comment by ptkato** [comment]
I believe there's some confusion regarding `v2-configure`, that command does not run the pre-build phase anymore, see #7411. However it might seem otherwise due to the fact that there's a `runRebuild` function amidst its code, but that's only to satisfy the typechecker, since `readProjectLocalExtraConfig` is wrapped around the `Rebuild` monad.

**Comment by fgaz** [comment]
Thanks for the contribution!

**Pull request merged**
